File: CP/LeetCode/Python/65.py
# 65. Valid Number

from __future__ import annotations


class Solution:
    def isNumber(self, s: str) -> bool:
        # Time complexity of regular expression is O(2^n)
        # Regular expression is not a good solution for this problem

        # import re
        # return bool(re.match(r'^[+-]?(\d+(\.\d*)?|\.\d+)([eE][+-]?\d+)?$', s.strip()))

        s = s.strip()

        seen_expont = seen_dot = seen_digit = False
        for i, c in enumerate(s):
            if c in "+-":
                if i > 0 and s[i - 1] not in "eE":
                    return False
            elif c in "eE":
                if seen_expont or not seen_digit:
                    return False
                seen_expont = True
                seen_digit = False
            elif c == ".":
                if seen_dot or seen_expont:
                    return False
                seen_dot = True
            elif c.isdigit():
                seen_digit = True
            else:
                return False

        return seen_digit


if __name__ == "__main__":
    sol = Solution()
    print(sol.isNumber("0"))  # True
    print(sol.isNumber(" 0.1 "))  # True
    print(sol.isNumber("abc"))  # False

File: CP/LeetCode/Java/65.java
// 65. Valid Number

class Solution {
    public boolean isNumber(final String s) {
        String newStr = new String(s);
        newStr = newStr.trim();
        if (newStr.length() == 0) {
            return false;
        }

        boolean isNumber = false;
        boolean isExponent = false;
        boolean isDecimal = false;

        for (int i = 0; i < newStr.length(); i++) {
            char ch = newStr.charAt(i);
            if (Character.isDigit(ch)) {
                isNumber = true;
            } else if (ch == '+' || ch == '-') {
                if (i > 0 && newStr.charAt(i - 1) != 'e') {
                    return false;
                }
            } else if (ch == 'e') {
                if (isExponent || !isNumber) {
                    return false;
                }
                isExponent = true;
                isNumber = false;
            } else if (ch == '.') {
                if (isDecimal || isExponent) {
                    return false;
                }
                isDecimal = true;
            } else {
                return false;
            }
        }

        return isNumber;
    }

    public static void main(final String[] args) {
        Solution solution = new Solution();
        System.out.println(solution.isNumber("0")); // true
        System.out.println(solution.isNumber(" 0.1 ")); // true
        System.out.println(solution.isNumber("abc")); // false
        System.out.println(solution.isNumber("1 a")); // false
        System.out.println(solution.isNumber("2e10")); // true
        System.out.println(solution.isNumber(" -90e3   ")); // true
        System.out.println(solution.isNumber(" 1e")); // false
        System.out.println(solution.isNumber("e3")); // false
        System.out.println(solution.isNumber(" 6e-1")); // true
        System.out.println(solution.isNumber(" 99e2.5 ")); // false
        System.out.println(solution.isNumber("53.5e93")); // true
        System.out.println(solution.isNumber(" --6 ")); // false
        System.out.println(solution.isNumber("-+3")); // false
        System.out.println(solution.isNumber("95a54e53")); // false
    }
}
